Add tracing to BLE GAP

noonfom opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->
The objective of the PR is to add traces to BLE GAP:
- [x] Complete GAP
- [x] Move string serialization to `ble_trace_helpers.h`
- [x] Add traces to PAL

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [ ] Feature update (New feature / Functionality change / New API)
    [ ] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [ ] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [ ] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->
@pan- 
@paul-szczepanek-arm 
<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

Commit:
Add traces to GapImpl.cpp

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1654,7 +1659,9 @@ ble_error_t Gap::createAdvertisingSet(
     const AdvertisingParameters &parameters
 )
 {
+    tr_info("attempting to create advertising set");
```
maybe instead put a trace at the end so you can print the handle of the set

pan- replied:
I think it is valuable to have a trace here as `setAdvertisingParameters` will probably generate more traces. Having a trace before explains why `setAdvertisingParameters` is being called. 
However, please add a trace with the handle allocated.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1610,13 +1613,15 @@ ble_error_t Gap::update_ll_address_resolution_setting()
 #if BLE_ROLE_BROADCASTER
 uint8_t Gap::getMaxAdvertisingSetNumber()
 {
+    tr_info("getting max. advertising set number");
```
instead print the trace when the set number is known

paul-szczepanek-arm replied:
I'm worried we call this function too often to log this every time

pan- replied:
I share that concern with you @paul-szczepanek-arm . My rule of thumb would be to not add log to getter excepts if it does lazy initialization (or start unexpected operations 😱). In that case, log the path of the operation but do not log the _getting_ part.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1905,13 +1936,15 @@ ble_error_t Gap::setAdvertisingData(
         if (handle == LEGACY_ADVERTISING_HANDLE) {
             prepare_legacy_advertising_set(AdvertisingParameters{});
         } else {
+            tr_error("handle must be equal to %d", LEGACY_ADVERTISING_HANDLE);
```
```suggestion
            tr_error("Trying to set advertising data on a non-existent set");
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1905,13 +1936,15 @@ ble_error_t Gap::setAdvertisingData(
 
 #if BLE_FEATURE_EXTENDED_ADVERTISING
     if (handle >= getMaxAdvertisingSetNumber()) {
+        tr_error("handle cannot be greater than/equal to %d", getMaxAdvertisingSetNumber());
```
tr_error("Trying to set advertising data on a non-existent set");

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -3339,6 +3373,9 @@ ble_error_t Gap::prepare_legacy_advertising_set(const AdvertisingParameters& par
 void Gap::setEventHandler(Gap::EventHandler *handler)
 {
+    if (handler == nullptr) {
+        tr_warning("setting event handler to a null pointer");
```
I know there's a group tag but gap has more than one handler
```suggestion
        tr_warning("setting Gap event handler to a null pointer");
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1924,6 +1957,7 @@ ble_error_t Gap::setAdvertisingData(
         }
 
         if (payload.size() > LEGACY_ADVERTISING_MAX_SIZE) {
+            tr_error("payload size cannot be greater than %d", LEGACY_ADVERTISING_MAX_SIZE);
```
```suggestion
            tr_error("Legacy advertising payload size cannot be greater than %d", LEGACY_ADVERTISING_MAX_SIZE);
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1739,18 +1755,22 @@ ble_error_t Gap::setAdvertisingParameters(
 #if BLE_FEATURE_PRIVACY
     // If privacy is enabled, alter the own address type used during advertising
     if (_privacy_enabled && params.getOwnAddressType() != own_address_type_t::RANDOM) {
+        tr_error("privacy is enabled, alter own address type");
```
```suggestion
        tr_error("Cannot use non random address in advertising set with privacy enabled.");
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1760,6 +1780,7 @@ ble_error_t Gap::setAdvertisingParameters(
         if (handle == LEGACY_ADVERTISING_HANDLE) {
             return prepare_legacy_advertising_set(params);
         } else {
+            tr_error("handle must be equal to %d", LEGACY_ADVERTISING_HANDLE);
```
```suggestion
            tr_error(""Trying to set advertising params on a non-existent set.");
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1694,33 +1702,41 @@ ble_error_t Gap::createAdvertisingSet(
     }
 
     if (handle == LEGACY_ADVERTISING_HANDLE) {
+        tr_error("handle cannot be equal to %d", LEGACY_ADVERTISING_HANDLE);
```
```suggestion
        tr_error("Legacy advertising set cannot be destroyed");
```

pan- replied:
I like the change @paul-szczepanek-arm but it is valuable to print the handle value.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1694,33 +1702,41 @@ ble_error_t Gap::createAdvertisingSet(
 #if BLE_FEATURE_EXTENDED_ADVERTISING
 ble_error_t Gap::destroyAdvertisingSet(advertising_handle_t handle)
 {
+    tr_info("destroying advertising set %d", handle);
```
I think it's better to trace the destruction when it's successful only

pan- replied:
If a user post a log for a bug, I'd prefer to have a log that allows me to understand what the application was doing. 
Info on operations started is very important for us. It also provide additional context for error traces.

paul-szczepanek-arm replied:
there are traces for all return values so all attempts will be logged

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1694,33 +1702,41 @@ ble_error_t Gap::createAdvertisingSet(
     }
 
     if (!_existing_sets.get(handle)) {
+        tr_error("advertising set does not exist");
```
```suggestion
        tr_error("Tried to destroy non-existent advertising set %d.", handle);
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1694,33 +1702,41 @@ ble_error_t Gap::createAdvertisingSet(
     }
 
     if (_pending_sets.get(handle) || _active_sets.get(handle)) {
+        tr_error("advertising set is pending/active");
```
```suggestion
        tr_error("Tried to destroy active advertising set %d.", handle);
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1694,33 +1702,41 @@ ble_error_t Gap::createAdvertisingSet(
     }
 #if BLE_FEATURE_PERIODIC_ADVERTISING
     if (_active_periodic_sets.get(handle)) {
+        tr_error("advertising set is active");
```
```suggestion
tr_error("Tried to destroy active periodic advertising set %d.", handle);
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1694,33 +1702,41 @@ ble_error_t Gap::createAdvertisingSet(
 
     ble_error_t err = _pal_gap.remove_advertising_set(handle);
     if (err) {
+        tr_error("PAL could not remove advertising set");
```
```suggestion
        tr_error("PAL could not remove advertising set %d", handle);
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1676,6 +1683,7 @@ ble_error_t Gap::createAdvertisingSet(
             );
 
             if (err) {
+                tr_error("failed to set advertising parameters");
```
```suggestion
                tr_error("Couldn't create advertising set - advertising parameters couldn't be set.");
```

Commit:
Add helpers and param logs

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -41,6 +47,94 @@ using namespace std::chrono;
 MBED_STATIC_ASSERT(BLE_GAP_MAX_ADVERTISING_SETS < 0xFF, "BLE_GAP_MAX_ADVERTISING_SETS must be less than 255");
 MBED_STATIC_ASSERT(BLE_GAP_MAX_ADVERTISING_SETS > 0, "BLE_GAP_MAX_ADVERTISING_SETS must be a positive number");
 
+const char* to_string(ble::advertising_type_t advType) {
```
these might be used also by other modueles, maybe better have them in some tracing header

pan- replied:
It will be once the SM PR is merged.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -41,6 +47,94 @@ using namespace std::chrono;
+            return "NON_CONNECTABLE_UNDIRECTED";
+        case ble::advertising_type_t::CONNECTABLE_DIRECTED_LOW_DUTY:
+            return "CONNECTABLE_DIRECTED_LOW_DUTY";
+        case ble::advertising_type_t::CONNECTABLE_NON_SCANNABLE_UNDIRECTED:
```
is this one missing on purpose?

pan- commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1654,7 +1748,9 @@ ble_error_t Gap::createAdvertisingSet(
     const AdvertisingParameters &parameters
 )
 {
+    tr_info("Create advertising set");
```
Can you detail the advertising parameters ?

pan- commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1739,18 +1845,60 @@ ble_error_t Gap::setAdvertisingParameters(
 {
+    tr_info("Set advertising parameters for set %d: "
+            "_advType=%s "
+            "_minInterval=%lums "
```
Can you use the macro constant (https://en.cppreference.com/w/c/types/integer) . `%lu` may not work well on all toulchain.

pan- commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1803,12 +1953,52 @@ ble_error_t Gap::setExtendedAdvertisingParameters(
 {
+    tr_info("Set extended advertising parameters for set %d: "
+            "_advType=%s "
+            "_minInterval=%lums "
```
Same as above.

pan- commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1768,6 +1917,7 @@ ble_error_t Gap::setAdvertisingParameters(
 #endif // BLE_FEATURE_EXTENDED_ADVERTISING
     {
         if (handle != LEGACY_ADVERTISING_HANDLE) {
+            tr_error("Failed: non-existent advertising set");
```
Bellow an error is not handled, you should add an else after the `if(!err)`

pan- commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1803,12 +1953,52 @@ ble_error_t Gap::setExtendedAdvertisingParameters(
         params.getUseLegacyPDU() == false
     ) {
         /* these types can only be used with legacy PDUs */
+        tr_error("Failed: directed and undirected connections can only be used with legacy PDUs");
```
Bellow an error is not handled, you should add an else after the `if(!err)`

pan- commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1856,6 +2046,10 @@ ble_error_t Gap::setAdvertisingPayload(
 )
 {
+    tr_info("Set advertising payload for set %d", handle);
+#if TRACE_WRITE_VALUES
```
I don't think we have to trace anything here, the trace can be in `setAdvertisingData` as this function just forward its call to it.

pan- commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1872,6 +2066,10 @@ ble_error_t Gap::setAdvertisingScanResponse(
     Span<const uint8_t> response
 )
 {
+    tr_info("Set advertising scan response for set %d", handle);
```
Same as `setAdvertisingPayload`.

pan- commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1889,6 +2087,10 @@ ble_error_t Gap::setAdvertisingData(
 )
 {
+    tr_info("Set advertising data for set %d", handle);
+#if TRACE_WRITE_VALUES
```
Why the payload is logged separatly ? The parameters minimise fragmentation and scan response are important too.

pan- commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1924,6 +2128,7 @@ ble_error_t Gap::setAdvertisingData(
         }
 
         if (payload.size() > LEGACY_ADVERTISING_MAX_SIZE) {
+            tr_error("Legacy advertising payload size cannot be greater than %d", LEGACY_ADVERTISING_MAX_SIZE);
```
Please add a trace instead of the `MBED_WARNING`

Commit:
Update GapImpl.cpp

Commit:
Move helpers to ble_trace_helpers.h

Commit:
Add traces to whitelist setter & getter functions

Commit:
Prepend parent dir. to ble_trace_helpers.h include line

Commit:
Resolve portability issue by using PRI* macro in place of %lu

Commit:
Add traces to PalGapImpl.cpp

Commit:
Complete PAL GAP traces

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/common/ble_trace_helpers.h:
```diff
@@ -200,6 +200,375 @@ static inline const char *to_string(privacy_mode_t mode)
     }
 }
 
+static inline const char* to_string(ble::advertising_type_t advType)
```
why would you want to inline these massive functions? especially if they're used in multiple places. The compiler will not deduplicate strings if they're in separate TUs

paul-szczepanek-arm replied:
Please add MBED_CONF_BLE_TRACE_HUMAN_READABLE_ENUMS to big ones

kjbracey replied:
Seems this was just copying the code above.

`static inline` is the worst of all worlds here - it prevents it from using common out-of-line definitions, forcing bloat whether it inlines or not, unless these are used in exactly one file.

You should only ever use `static inline` in headers for super-super-tiny stuff you're confident will be inlined. (And even that can cause problems - it means such functions can't be called from non-static `inline` functions). Only strong reason I've seen for `static inline` is C99/C++ header compatibility, but it seems that non-static `inline` interoperates fine in mixed C99/C++ these days, as long as you remember to provide C99 `extern` definitions in a C file.

You definitely need to take the `static` off all these.

A couple of the smallest functions above with just `if`/`else` could just possibly be an inlining win, but I doubt it. So I'd agree the `inline` is a bit pointless, but not that bad - the compiler can and probably will just choose not to inline.

Main potential benefit of `inline` is that it would permit constant optimisation. I doubt any of these would get passed constants often though.

PS Linkers can deduplicate strings, unless they're placed alongside the code, but I think only ARMC5 likes to do that. In ARMC5 having these non-inline would permit them to do local sharing of `"PUBLIC"`, `"unknown"` and some others, if `--split-sections` was turned off.

kjbracey replied:
I'm talking about the new functions and the existing ones, yes.

None of this header code should be `static inline`. Conceivably the `if/else` ones might benefit from simple `inline`, and some of the other little `mbed_trace_array` ones. I doubt any of the others do unless they're ever invoked with a constant. But simple `inline` shouldn't be harmful to code size - main objection would be on grounds of style or compilation time.

Just stripping all the `static` off would satisfy me, but maybe not Paul.

It's possible that the reason for the flood of `static inline` here is actually just a cascade effect from some inner function being `static inline`. Once you've done that, you can't call it from a plain `inline` function. (Because that would then be a common definition calling a translation-unit-local definition).

That's bitten me with CMSIS `__WFE()` type functions. They're `static inline`, which doesn't pose a bloat issue cos they're tiny and always inlined, but you can't then use them in `inline` functions.

paul-szczepanek-arm replied:
I think this all just stemmed from some initial assumption that maybe was true for the first one (some function somewhere that was converting a compile time values in a couple of places) and just got out of hand. These are used from multiples places so as long as static is gone I'm fine with it, I'm sure the compiler will ignore inline anyway.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/PalGapImpl.cpp:
```diff
@@ -116,6 +126,24 @@ ble_error_t PalGap::set_advertising_parameters(
     advertising_filter_policy_t advertising_filter_policy
 )
 {
+    tr_info("[PAL] Set advertising parameters - "
```
this is already traced in gap, I don't think there's any benefit of tracing it twice, especially since it's so long

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/PalGapImpl.cpp:
```diff
@@ -157,6 +185,12 @@ ble_error_t PalGap::set_advertising_data(
     const advertising_data_t &advertising_data
 )
 {
+    tr_info("[PAL] Set advertising data - "
```
already traced

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/PalGapImpl.cpp:
```diff
@@ -173,6 +207,12 @@ ble_error_t PalGap::set_scan_response_data(
     const advertising_data_t &scan_response_data
 )
 {
+    tr_info("[PAL] Set scan response data - "
```
please go through these and remove duplicate traces

paul-szczepanek-arm replied:
if there is a meaningful conversion or parameter checking - you can trace that instead.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/common/ble_trace_helpers.h:
```diff
@@ -200,6 +200,375 @@ static inline const char *to_string(privacy_mode_t mode)
+
+static inline const char* to_string(ble::advertising_channel_map_t map)
+{
+    switch (map.value()) {
```
it's a 3 bit field, there must be a better way of tracing it than printing the long enum names,
"[101]" is probably even more readable

paul-szczepanek-arm reviewed (changes requested):
review in progress

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/PalGapImpl.cpp:
```diff
@@ -244,6 +289,18 @@ ble_error_t PalGap::set_scan_parameters(
     scanning_filter_policy_t filter_policy
 )
 {
+    tr_info("[PAL] Set scan parameters -"
```
missing space after `-` but this whole trace is duplicate anyway

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/PalGapImpl.cpp:
```diff
@@ -472,11 +615,13 @@ ble_error_t PalGap::disconnect(
         DmReadPhy(connection);
         return BLE_ERROR_NONE;
     }
+    tr_error("[PAL] Failed: LE_2M_PHY feature not supported");
```
or coded
```suggestion
    tr_warning("[PAL] Failed: Only 1M PHY supported by controller");
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/PalGapImpl.cpp:
```diff
@@ -1424,6 +1744,18 @@ ble_error_t PalGap::extended_create_connection(
+    tr_info("[PAL] Extended create connection - "
+            "initiator_policy=%s, "
+            "own_address_type=%s, "
+            "peer_address_type=%s,"
```
missing space, but I think the whole trace is duplicate from gap

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -19,6 +19,9 @@
 #include <cstdint>
 
+// REMOVE ONCE MOVED TO COMMON FILE
+#include <stdlib.h>
```
?

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -522,17 +537,51 @@ ble_error_t Gap::connect(
+
+    // Trace connection params for each enabled PHY
+    for (size_t i = 0; i < connectionParams.getPhySet().count(); ++i) {
+        tr_info("PHY %d - "
```
maybe some parenthesis would help show where one starts and ends

paul-szczepanek-arm replied:
good point, disregard my comment then

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -668,6 +722,8 @@ ble_error_t Gap::connect(
     if (ret == BLE_ERROR_NONE) {
         _initiating = true;
+    } else {
+        tr_error("Failed: PAL could not create connection");
```
there's a ble_error_to_string() you can use now

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -3253,16 +3611,27 @@ ble_error_t Gap::removeDeviceFromPeriodicAdvertiserList(
     }
 
     if (sid > 0x0F) {
+        tr_error("Failed: sid cannot be greater than 0x0F");
```
```suggestion
        tr_error("Failed: Set ID cannot be greater than 0x0F");
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -3135,17 +3458,32 @@ ble_error_t Gap::createSync(
     if (peerAddressType != peer_address_type_t::PUBLIC &&
         peerAddressType != peer_address_type_t::RANDOM
         ) {
+        tr_error("Failed: peer address type must be public/random");
```
There is a lot of variation on this message, probably 20 different strings, they could all be the same string:
```suggestion
        tr_error("can't use extended feature - controller doesn't support extended advertising");
```

paul-szczepanek-arm replied:
This comment is on the wrong line - sorry, no idea how this happened. This comment is meant for times when you check if extended advertising is present. This line is fine, yes.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -3043,7 +3354,16 @@ ble_error_t Gap::startScan(
+            period.valueInMs());
+
     if (_initiating) {
+        tr_error("Failed: stack busy");
```
```suggestion
        tr_error("Cannot start scan - busy trying to connect");
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -2164,6 +2423,7 @@ ble_error_t Gap::stopAdvertising(advertising_handle_t handle)
 bool Gap::isAdvertisingActive(advertising_handle_t handle)
 {
     if (handle >= getMaxAdvertisingSetNumber()) {
+        tr_error("Cannot determine if periodic advertising is active: handle greater than/equal to %d", getMaxAdvertisingSetNumber());
```
I don't think this is specific to periodic, also a simple "invalid advertising handle %d" would be enough

paul-szczepanek-arm replied:
What I meant is: this is not for periodic advertising, this is normal advertising.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -2108,19 +2362,23 @@ ble_error_t Gap::startAdvertising(
 #endif // BLE_FEATURE_EXTENDED_ADVERTISING
 
     if (!_active_sets.get(handle) || _pending_sets.get(handle)) {
+        tr_error("Failed: active advertising set");
```
```suggestion
        tr_error("Failed: Cannot stop set - not active");
```

paul-szczepanek-arm reviewed (changes requested):
I don't think we need the word "Failed: " at the start of each error, it's already an error.

We should either remove the duplicate traces from PAL or at the very least have the PAL tracing as a separate config option. These traces are almost identical.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/PalGapImpl.cpp:
```diff
@@ -887,15 +1052,62 @@ ble_error_t PalGap::set_extended_advertising_parameters(
+            secondary_advertising_max_skip,
+            to_string(secondary_phy),
+            advertising_sid,
+            to_string(scan_request_notification));
```
```suggestion
    tr_info("[PAL] Advertising set %d: set extended advertising parameters - "
            "event_properties:[%s%s%s%s%s%s%s ], "
            "primary_advertising_interval_min=%" PRIu32 ", "
            "primary_advertising_interval_max=%" PRIu32 ", "
            "primary_advertising_channel_map=%s, "
            "own_address_type=%s, "
            "peer_address_type=%s, "
            "peer_address=%s, "
            "advertising_filter_policy=%s, "
            "advertising_power=%d, "
            "primary_advertising_phy=%s, "
            "secondary_advertising_max_skip=%d, "
            "secondary_phy=%s, "
            "advertising_sid=%d, "
            "scan_request_notification=%s",
            advertising_handle,
            event_properties.connectable ? " connectable" : "",
            event_properties.scannable ? " scannable" : "",
            event_properties.directed ? " directed" : "",
            event_properties.high_duty_cycle ? " high_duty_cycle" : "",
            event_properties.use_legacy_pdu ? " use_legacy_pdu" : "",
            event_properties.omit_advertiser_address ? " omit_advertiser_address" : "",
            event_properties.include_tx_power ? " include_tx_power" : "",
            primary_advertising_interval_min,
            primary_advertising_interval_max,
            to_string(primary_advertising_channel_map),
            to_string(own_address_type),
            to_string(peer_address_type),
            to_string(peer_address),
            to_string(advertising_filter_policy),
            advertising_power,
            to_string(primary_advertising_phy),
            secondary_advertising_max_skip,
            to_string(secondary_phy),
            advertising_sid,
            to_string(scan_request_notification));
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -2985,7 +3278,24 @@ void Gap::on_remote_connection_parameter(
+            "own_address_type=%s, "
+            "scanning_filter_policy=%s, "
+            "phys=%d ,"
+            "phy_1m_configuration:[interval=%lums,window=%lums,active_scanning=%s] ,"
```
```suggestion
            "phy_1m_configuration:[interval=%lums,window=%lums,active_scanning=%s], "
```

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -2985,7 +3278,24 @@ void Gap::on_remote_connection_parameter(
+    tr_info("Set scan parameters - "
+            "own_address_type=%s, "
+            "scanning_filter_policy=%s, "
+            "phys=%d ,"
```
```suggestion
            "phys=%d, "
```

paul-szczepanek-arm replied:
maybe instead of printing this enum as integer just print the phys that are being configured
if (phy1) tr_info("phy1 config")
if (phy coded) tr_info("coded config")

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -2985,7 +3278,24 @@ void Gap::on_remote_connection_parameter(
+            to_string(params.get1mPhyConfiguration().isActiveScanningSet()),
+            params.getCodedPhyConfiguration().getInterval().valueInMs(),
+            params.getCodedPhyConfiguration().getWindow().valueInMs(),
+            to_string(params.getCodedPhyConfiguration().isActiveScanningSet()));
```
should be dependent on the phy selected

paul-szczepanek-arm replied:
why are you printing coded phy configuration if it's not set?

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -3205,6 +3551,7 @@ ble_error_t Gap::cancelCreateSync()
 ble_error_t Gap::terminateSync(periodic_sync_handle_t handle)
 {
```
maybe worth adding the tr_info for the user call

Commit:
Remove stray include

Commit:
Make 'echo' traces in PAL configurable

Some of the tr_info traces at the beginning of PAL functions  are duplicates ('echoes') of those in GAP functions. This commit enables including/excluding of these traces in mbed_app.json by writing 1/0, respectively to the MBED_CONF_CORDIO_TRACE_PAL_ECHOES config macro. The default is to exclude 'echo' traces.

Commit:
Add missing spaces in tr_info traces

pan- commented on connectivity/FEATURE_BLE/source/cordio/source/PalGapImpl.cpp:
```diff
@@ -186,6 +231,10 @@ ble_error_t PalGap::set_scan_response_data(
 
 ble_error_t PalGap::advertising_enable(bool enable)
 {
+    tr_info("[PAL] Advertising enable - "
```
We probably don't need the `[PAL]` prefix in this file. I would prefer to use a different `TRACE_GROUP` that indicates this is the PAL. `BLDM` ?

Commit:
Clean-up param log for set_extended_advertising_parameters

pan- commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -976,12 +1093,18 @@ ble_error_t Gap::setWhitelist(const whitelist_t &whitelist)
 #if BLE_FEATURE_PRIVACY
 ble_error_t Gap::enablePrivacy(bool enable)
 {
+    tr_info("Enable privacy -"
```
Could you add traces bellow to `setPeripheralPrivacyConfiguration` ?

Commit:
Remove "Failed: " from every tr_error trace

Commit:
Apply changes from PalGapImpl.cpp review

Commit:
Apply changes from GapImpl.cpp review

Commit:
Use binary for channel map trace

B0: ADVERTISING_CHANNEL_37
B1: ADVERTISING_CHANNEL_38
B2: ADVERTISING_CHANNEL_39

We're assuming that the user understands the above. This could be documented somewhere.

Commit:
Replace more %lu w/PRIu32

Commit:
Add tr_info to terminateSync for user call

pan- reviewed (dismissed):
Thanks for the updates @noonfom. The calling path is quite comprehensive now. However there is not a lot of information logged from events whether it is in the PAL or the GapImplementation.

It is probably not necessary to add these logs to the PAL file as it is directly calling the generic gap handlers. However, I'd like to see logs of event handling in the GapImpl.cpp file so we know when a connection has been established for example. For reccuring events such as advertising received or scan request received, you probably want to use the debug level to not saturate the system.

Commit:
Make trace group for PAL traces BLDM

Commit:
Remove [PAL] prefix from every PAL trace

Commit:
Add trace to setPeripheralPrivacyConfiguration

Commit:
Make set scan params tr_info dependent on selected PHY

Commit:
Add trace to GAP events

pan- commented on connectivity/FEATURE_BLE/source/common/ble_trace_helpers.h:
```diff
@@ -214,6 +215,439 @@ static inline const char *to_string(privacy_mode_t mode)
+        case ble::advertising_channel_map_t::ADVERTISING_CHANNEL_38:
+            return "[010]";
+        case ble::advertising_channel_map_t::ADVERTISING_CHANNEL_37_AND_38:
+            return "[011";
```
There is a missing `]`

pan- commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -522,17 +534,51 @@ ble_error_t Gap::connect(
+            to_string(connectionParams.getOwnAddressType()));
+
+    // Trace connection params for each enabled PHY
+    for (size_t i = 0; i < connectionParams.getPhySet().count(); ++i) {
```
Could you add a trace in `on_unnexpected_error` ?

pan- replied:
Same source file, wrong spelling from me: `on_unexpected_error`

pan- commented on connectivity/FEATURE_BLE/source/cordio/source/PalGapImpl.cpp:
```diff
@@ -1424,6 +1770,20 @@ ble_error_t PalGap::extended_create_connection(
 )
 {
+#if MBED_CONF_CORDIO_TRACE_PAL_ECHOES
+    tr_info("Extended create connection - "
```
Could you add traces below in `update_direct_advertising_parameters` ? There is error paths which are not covered.

pan- commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -1174,6 +1309,8 @@ void Gap::on_scan_started(bool success)
 
 void Gap::on_scan_stopped(bool success)
 {
+    tr_info("Scan %s", success ? "successfully stopped" : "failed to stop");
```
I think we can add traces bellow explaining that the address has been refreshed and advertising will restart.

pan- commented on connectivity/FEATURE_BLE/source/generic/GapImpl.cpp:
```diff
@@ -864,11 +973,15 @@ uint8_t Gap::getMaxWhitelistSize(void) const
 
 ble_error_t Gap::getWhitelist(whitelist_t &whitelist) const
 {
+    tr_info("Get whitelist");
```
Do we need a trace to indicate we are using this getter ? The trace in the error path should be sufficient if it mentions that it fails while trying to get the whitelist. However we do need traces in `initialize_whitelist`.

pan- reviewed (dismissed):
Thanks for adding more relevant traces, I believe some are still missing: 
- `on_extended_advertising_report` (tr_debug)
- More detail about the connection process, for example, we disconnect in `signal_connection_complete` if the system is low on resources. Or when the connection is dismissed due to policy settings. 
- `on_periodic_advertising_sync_established`
- `on_periodic_advertising_report` (tr_debug)
- `on_periodic_advertising_sync_loss`
- `on_advertising_set_started`
- `on_legacy_advertising_started`
- `on_legacy_advertising_stopped`
- `on_advertising_set_terminated`
- `on_scan_request_received`
- `on_connection_update_complete`
- `on_remote_connection_parameter`
- `on_private_address_generated` (we should detail what is hapenning)
- `on_address_resolution_completed` tr_debug for the advertising path and `tr_info` for the connection path could be valuable. 
-

Commit:
Other GAP events required traces

pan- reviewed (approved):
Thanks, I think we are mostly there. I'd like to merge this PR. We still have some work to do to merge that branch back to master. During that phase, we will fix any issue detected.
